src/agent/internals/src/main/java/com/google/devtools/cdbg/debuglets/java/FirebaseClient.java

```diff
@@ -0,0 +1,843 @@
+   * listActiveBreakpoints() needs to retrieve and return to the native agent code.
+   *
+   * <p>It is expressly created with a size of 1, as it only ever needs the most recent snapshot,
+   * older unprocessed snapshots can be purged when a new snapshot arrives.
+   *
+   * <p>There is one reader and one writer. The writer is the breakpoints update callback that is
+   * registered with the Firebase Admin SDK, and runs whenever the active breakpoints snapshot on
+   * the Firebase RTDB changes. Everytime it runs, it will clear out the current entry (if any), and
+   * add in the latest snapshot. The reader the native agent code which calls
+   * listActiveBreakpoints() to stay up to date on the current snapshot it should have installed.
+   *
+   * <p>The key is the breakpoint ID, and the value is a POJO that can be naturally serialized as a
+   * JSON document. In practice since it represents a Breakpoint, the Object itself will be a
+   * Map<String, Object> instance.
+   */
+  private ArrayBlockingQueue<Map<String, Object>> pendingActiveBreakpoints =
+      new ArrayBlockingQueue<>(1);
+
+  /**
+   * Holds the current set of Active Breakpoints. It is used when the native agent code calls
+   * transmitBreakpointUpdate(), the breakpoint is retrieved from here to fill in the gaps of the
+   * breakpoint as we send the entire Breakpoint data tree to be update the node in the backing
+   * Firebase RTDB. This 'filling in' is needed because the breakpoint data that comes up from the
+   * native code does not contain all of the fields of the breakpoint, and so this ensures no
+   * breakpoint fields are lost. This map also holds any modifications to the breakpoint the native
+   * agent code may have made, such as an updated line location, since this data may not yet be in
+   * the Firebase RTDB and we need to ensure we don't accidentally lose this local change.
+   *
+   * <p>There are two writers, and one reader. One writer is the breakpoints update callback that is
+   * registered with the Firebase Admin SDK, and runs whenever the active breakpoints snapshot on
+   * the Firebase RTDB changes. It will add new active breakpoints, and it will remove breakpoints
+   * that are no longer active. However it will not update any breakpoint entries that are still
+   * active, as they may contain a locally updated breakpoint location. The other writer, which is
+   * also the reader, is the native agent code via its call to transmitBreakpointUpdate(). If the
+   * breakpoint update being transmitted is not finalizing a breakpoint (eg. it's a source line
+   * update), then it updates the particular breakpoint entry. It also acts as a reader when it
+   * needs to combine the breakpoint fields provided from the native agent code with the complete
+   * breakpoint to be written to the Firebase RTDB.
+   */
+  private ConcurrentHashMap<String, Object> currentBreakpoints = new ConcurrentHashMap<>();
```

Comment:
I'm not sure how much of this is my appreciation of Go conventions, but it would be nice if we had some slightly more descriptive types.  BreakpointId instead of String where appropriate, and possibly even BreakpointData instead of map<String, Object>.  The current types make it harder for me to read the code.

Reply:
I'm not too sure on the Go side, but if we had at least a typedef like in C++ that would help.  In Java, the main way I can think of would be to wrap the types, and then write all the necessary forwarding methods, but that doesn't seem ideal either. Or is that what you were thinking?